### PR TITLE
[SOLUTION: NXLog DNSLogs 3.0.0]

### PR DESCRIPTION
**Pull request by jszigetvari-nxlog**
This PR addresses the feedback received from Microsoft regarding broken links

CHANGES:

- Corrected outdated NXLog documentation and support URLs
- Updated Package 2.0.1.zip to 2.0.2.zip
- Updated and extended the ASimDnsMicrosoftNXLog parser

REASONS FOR CHANGE(S):

- To address the findings raised by Microsoft for this and other solutions
- I took over the maintenance tasks from former colleagues who worked on this earlier
- The Azure Sentinel repo maintainers updated the in-repo version of the parser, and I mostly applied those changes, and extended some data enrichment code

VERSION UPDATED:

- Yes, to version 2.0.2

TESTING COMPLETED:

- Need Help / I did not do any actual tests on this, but compared to the in-repo data created by my former colleagues, I made made no breaking changes

RAN VALIDATIONS:

- Yes, all validations passed successfully

This work is associated with NXLog's Jira Issue IN-270

### CODE REVIEW

**Comment by jszigetvari-nxlog** [comment]
@v-rbajaj @devikamehra is there perhaps something that I could to help you with the review process?

**Comment by devikamehra** [comment]
@v-rbajaj I will review the ASim Parser. Are all the other components verified?

**Comment by v-rbajaj** [comment]
Hi @devikamehra, yes, verified all the other components, they are working fine.

**Comment by devikamehra** [comment]
@jszigetvari-nxlog There are fundamental flaws in the execution of this solution. Parser folder with the folder should only contain Source specific parsers and not ASIM parsers. ASIM Parser lies in [this](https://github.com/Azure/Azure-Sentinel/tree/master/Parsers) Parser folder under the required schema folder.

Moreover, the format of the parser is not correctly mapped with standard ASIM parsers. The parser should be in YAML in specific format. Please refer to any one ASIM parser to know more.

ASIM parsers do not have any filters (parameters) attached to them. However, the VIM version of the ASIM parser do have these parameter. Please check ASIM development guideline to know more.

Also, NXLog ASIM parsers (both with and without parameters) do already exist. How is the parser different from the once already present. If you find there is anything missing, please raise a separate PR to include those changes instead of adding a new one in the solution.

@v-rbajaj Can we check how we allowed an ASIM parser within the solution instead of adding it in Parser folder?

**Comment by jszigetvari-nxlog** [comment]
@devikamehra Please find my replies in-line

> There are fundamental flaws in the execution of this solution. Parser folder with the folder should only contain Source specific parsers and not ASIM parsers. ASIM Parser lies in [this](https://github.com/Azure/Azure-Sentinel/tree/master/Parsers) Parser folder under the required schema folder.

The last modification to this file happened almost 2 years ago by a colleage who has since left the company. I was asked to update parts of the NXLog DNS solution (along with others) that needed updating, like dead links and such.
I just took what was already done and update and modify existing things to get familiar with the contribution process.

> ASIM parsers do not have any filters (parameters) attached to them. However, the VIM version of the ASIM parser do have these parameter. Please check ASIM development guideline to know more.

The parser you mention in the /Parsers/ASimDns/Parsers folder was created by two Microsoft employees oshezaf and yaronMSFT, although the ASIM parser included in this solution may have been a model for that.

> Also, NXLog ASIM parsers (both with and without parameters) do already exist. How is the parser different from the once already present. If you find there is anything missing, please raise a separate PR to include those changes instead of adding a new one in the solution.

I just assumed the ASIM parser included in the solution would be installed and made available when the solution is deployed. Isn't that the case? Did it ever work?


So how should I now proceed? Shall I remove the ASIM parser from the solution completely and submit a new PR with my changes to the parser(s) separately?

Also, I noticed there are four versions of the parser:
Parsers/ASimDns/ARM/vimDnsMicrosoftNXlog/vimDnsMicrosoftNXlog.json, Parsers/ASimDns/ARM/ASimDnsMicrosoftNXlog/ASimDnsMicrosoftNXlog.json, Parsers/ASimDns/Parsers/vimDnsMicrosoftNXlog.yaml and Parsers/ASimDns/Parsers/ASimDnsMicrosoftNXlog.yaml

Which ones should I modify? All of them?
The json ones (in the ARM directory) seem to be more aimed at deployment and are less human-friendly. Are they somehow generated from the yaml variants or am I supposed to edit those files by hand too?
What is the difference between the vim and ASim parser variants? If I understand right, the vim versions have parameters, while the ASim ones do not. Is that the only difference?

Thanks for your help!

**Comment by jszigetvari-nxlog** [comment]
@v-rbajaj , @devikamehra Could you please help me in addressing the latest findings?

**Comment by devikamehra** [comment]
@v-rbajaj Please ensure that the solution is in the right shape. Replace the name of the Parser in this folder and change it accordingly in schema file as well as data connector. You may want to connect with Prateek to know more about the standard process.
@jszigetvari-nxlog In case you wish to make your contribution in ASim Parser, please modify the files present in Parsers/ASimDns/Parsers/ only.
ARM templates will get auto-generated once you check-in your code in the "Parsers/ASimDns/Parsers/" folder. No need to update them manually.
Yes, vim and asim parsers are different based on the usage of parameters only.

**Comment by jszigetvari-nxlog** [comment]
@devikamehra Thank you for your answer!
@v-rbajaj 
What I have planned as a next step is to remove the Parser folder for this solution completely, update the `mainTemplate.json` and `createUiDefinition.json` files, not to list the parser in question.

On the other hand, it turned out that the data connector included in this soliton relies on the `ASimDnsMicrosoftNXLog` parser:

```    "sampleQueries": [
        {
            "description" : "DNS Server top 5 hostlookups",
            "query": "ASimDnsMicrosoftNXLog \n| summarize count() by Domain\n| take 5\n| render piechart title='Top 5 host lookups'"
        },
        {
            "description" : "DNS Server Top 5 EventOriginalTypes (Event IDs)",
            "query" : "ASimDnsMicrosoftNXLog \n| extend EventID=strcat('Event ID ',trim_end('.0',tostring(EventOriginalType)))\n| summarize CountByEventID=count() by EventID\n| sort by CountByEventID\n| take 5\n| render piechart title='Top 5 EventOriginalTypes (Event IDs)'"
        },
        {
            "description" : "DNS Server analytical events per second (EPS)",
            "query" : "ASimDnsMicrosoftNXLog \n| where EventEndTime >= todatetime('2021-09-17 03:07')\n| where EventEndTime <  todatetime('2021-09-18 03:14')\n| summarize EPS=count() by bin(EventEndTime, 1s)\n| render timechart title='DNS analytical events per second (EPS) - All event types'"
        }
    ],```

How could I make sure that the official `ASimDnsMicrosoftNXLog` parser ends up getting used?

Also, will the changes to `.script/tests/KqlvalidationsTests/CustomFunctions/ASimDnsMicrosoftNXLog.json` be still needed if I remove the parser included in the solution? I believe it should be safe to remove that file afterwards.

**Comment by devikamehra** [comment]
@jszigetvari-nxlog There is no action item on you for changes in the solution. @v-rbajaj Should be taking those up.

In case you want to make any changes in the ASIM parser, please raise a separate PR for the same.

**Comment by v-rbajaj** [comment]
Hi @devikamehra and @jszigetvari-nxlog, I am checking on the folder structure. Will get in touch with Prateek, Sorry for the delay in response here.

**Comment by jszigetvari-nxlog** [comment]
@v-rbajaj Thanks for the feedback and setting the ETA.

**Comment by ashishsyal** [comment]
We are working on this PR - will share the status by EOD.

**Comment by v-rbajaj** [comment]
Hi @devikamehra and @jszigetvari-nxlog, apologies for delay in response. We are not able to push changes on this branch since it is a protected branch. 
We would either need to merge these changes and raise a new PR with the requested changes from our branch.

**Comment by jszigetvari-nxlog** [comment]
Dear @v-rbajaj  could you please try to push your changes to the following branch of my repo?
https://github.com/jszigetvari-nxlog/Azure-Sentinel/tree/nxlog-dns-solution-2.0.2
This PR is based on that branch, and I believe if you push your changes there, you may be able to add them to this PR.
Please let me know if I need to do anything to ensure that you can push your changes to that branch.

**Comment by v-rbajaj** [comment]
We will later look into how are we going to push the changes for this PR. Right now working on additional changes which are required for this PR. Will provide next update by 20th Jul,2023.

**Comment by v-rbajaj** [comment]
Hi @jszigetvari-nxlog, we have created a new branch on top of this branch, made the required changes and created a new PR
Will keep updating this PR --> https://github.com/Azure/Azure-Sentinel/pull/8581

**Comment by jszigetvari-nxlog** [comment]
Thanks @v-rbajaj for the update, I will check and comment your PR. I think we should keep this PR open, until we merge #8581

**Comment by v-rbajaj** [comment]
Sure, thanks @jszigetvari-nxlog.

**Review by vakohl — APPROVED** [review]

**Review by vakohl — CHANGES_REQUESTED** [review]
Approved by mistake, review is pending

**Comment by v-rbajaj** [comment]
Hi @jszigetvari-nxlog, I have raised a new PR for the parser changes and solution changes, ill update the PR#8722 with more changes.

**Comment by jszigetvari-nxlog** [comment]
@v-rbajaj Thanks for letting me know. I will monitor #8722 too.

**Comment by vakohl** [comment]
ETA for this review is 18th Aug

**Comment by jszigetvari-nxlog** [comment]
@vakohl It has been determined, that the structure of this solution is not ideal. This PR should stay open until the two related subtasks are done:
#8722 The parser should be removed from the solution, and the rest of the solution should be put into shape for the merging ( @v-rbajaj )
#8516 The changes to the ASim parser should be validated and merged ( @vakohl )

After those are complete and merged, this PR can be closed.

**Comment by vakohl** [comment]
> @vakohl It has been determined, that the structure of this solution is not ideal. This PR should stay open until the two related subtasks are done: #8722 The parser should be removed from the solution, and the rest of the solution should be put into shape for the merging ( @v-rbajaj ) #8516 The changes to the ASim parser should be validated and merged ( @vakohl )
> 
> After those are complete and merged, this PR can be closed without merging.

Provided comments, please have a look https://github.com/Azure/Azure-Sentinel/pull/8516/files

File: Solutions/NXLogDnsLogs/Parsers/ASimDnsMicrosoftNXLog.txt
Comment (vakohl) [inline Solutions/NXLogDnsLogs/Parsers/ASimDnsMicrosoftNXLog.txt:10]: @jszigetvari-nxlog  thankyou for your suggestion regarding modifying this solution parser.  We can use this parser to query ASIM function. Please replace this parser with the following query:

_Im_Dns
| where EventVendor == "Microsoft" and EventProduct == "DNS Server"

**Comment by jszigetvari-nxlog** [comment]
@vakohl There were some changes in the upstream repo, so I had to sync up with those.
I also tried to package the solution with the v3 packaging script, but got the following error:

```
    [+] Dynamic Variable References Should Not Use Concat (6 ms)
  IDs Should Be Derived From ResourceIDs
    [-] IDs Should Be Derived From ResourceIDs (127 ms)
        Property: "contentProductId" must use one of the following expressions for an resourceId property:
            extensionResourceId,resourceId,subscriptionResourceId,tenantResourceId,if,parameters,reference,variables,subscription,guid
        Property: "id" must use one of the following expressions for an resourceId property:
            extensionResourceId,resourceId,subscriptionResourceId,tenantResourceId,if,parameters,reference,variables,subscription,guid
        Property: "contentProductId" must use one of the following expressions for an resourceId property:
            extensionResourceId,resourceId,subscriptionResourceId,tenantResourceId,if,parameters,reference,variables,subscription,guid
        Property: "id" must use one of the following expressions for an resourceId property:
            extensionResourceId,resourceId,subscriptionResourceId,tenantResourceId,if,parameters,reference,variables,subscription,guid
        Property: "contentProductId" must use one of the following expressions for an resourceId property:
            extensionResourceId,resourceId,subscriptionResourceId,tenantResourceId,if,parameters,reference,variables,subscription,guid
        Property: "id" must use one of the following expressions for an resourceId property:
            extensionResourceId,resourceId,subscriptionResourceId,tenantResourceId,if,parameters,reference,variables,subscription,guid

  Location Should Not Be Hardcoded
    [+] Location Should Not Be Hardcoded (171 ms)
```

I'm also running the standard tests...
I'm also puzzled a bit, because I have no `contentProductId` field defined anywhere, `id` I may have here and there, but it's not clear which one I should correct.
Could you help me a bit with this please?

**Comment by jszigetvari-nxlog** [comment]
@vakohl If using the v3 packaging, should I update the package version to 3.0.0?

**Comment by v-atulyadav** [comment]
Hi @jszigetvari-nxlog, you can proceed with package version 3.0.0.

**Comment by jszigetvari-nxlog** [comment]
UPDATE OF DESCRIPTION:

This PR addresses the feedback received from Microsoft regarding broken links

CHANGES:

    Corrected outdated NXLog documentation and support URLs
    Updated Package 2.0.1.zip to 3.0.0.zip
    Updated the ASimDnsMicrosoftNXLog parser to fall back to using the official ASimDns parser
      in order to avoid duplicating code

REASONS FOR CHANGE(S):

    To address the findings raised by Microsoft for this and other solutions
    I took over the maintenance tasks from former colleagues who worked on this earlier
    The Azure Sentinel repo maintainers updated the in-repo version of the parser, and I mostly applied those changes, and extended some data enrichment code

VERSION UPDATED:

    Yes, to version 3.0.0

TESTING COMPLETED:

    Tests are passing OK

RAN VALIDATIONS:

    Yes, all validations passed successfully

This work is associated with NXLog's Jira Issue IN-270 and IN-451

**Comment by jszigetvari-nxlog** [comment]
@v-atulyadav Thanks for your help!
@vakohl All the changes requested, have been applied.

File: Sample Data/Custom/DNS_Logs_CL.json
Comment (vakohl) [inline Sample Data/Custom/DNS_Logs_CL.json:1]: why we are removing this?

File: Solutions/NXLogDnsLogs/Data Connectors/NXLogDnsLogs.json
Comment (vakohl) [inline Solutions/NXLogDnsLogs/Data Connectors/NXLogDnsLogs.json:2]: @v-rbajaj Parser looks good to me, can you please have a look at the other changes once?

**Review by vakohl — CHANGES_REQUESTED** [review]

File: Solutions/NXLogDnsLogs/Parsers/ASimDnsMicrosoftNXLog.txt
Comment (vakohl) [reply Solutions/NXLogDnsLogs/Parsers/ASimDnsMicrosoftNXLog.txt]: Looks good

File: Sample Data/Custom/DNS_Logs_CL.json
Comment (jszigetvari-nxlog) [reply Sample Data/Custom/DNS_Logs_CL.json:1]: I removed that because it was obsolete.
Throughout the solution we use the table called `NXLog_DNS_Server_CL` and for some time any references to `DNS_Logs_CL` have been removed. That old sample file was introduced back in #1280 
For `NXLog_DNS_Server_CL` we have sample logs already, and since `DNS_Logs_CL` is not used anywhere anymore, IMHO there is no need for any sample logs for it.

Comment (vakohl) [reply Sample Data/Custom/DNS_Logs_CL.json:1]: @jszigetvari-nxlog lets not remove this file. This file seems to have sample data relevant for NXLog too. Let's keep it as is.

Comment (jszigetvari-nxlog) [reply Sample Data/Custom/DNS_Logs_CL.json:1]: @vakohl I believe I can safely say that it is not used or referenced anywhere from our side. But if you'd like, I can put it back.

Comment (vakohl) [reply Sample Data/Custom/DNS_Logs_CL.json:1]: Sample logs are not referenced directly, these are useful for future requirements like troubleshooting or verifying parser. If incase there any need comes where we have to explore different available logs, these would be useful. I would keep it as is.

Comment (jszigetvari-nxlog) [reply Sample Data/Custom/DNS_Logs_CL.json]: All right, I have re-added that file.

File: .script/tests/KqlvalidationsTests/CustomFunctions/ASimDnsMicrosoftNXLog.json
Comment (vakohl) [inline .script/tests/KqlvalidationsTests/CustomFunctions/ASimDnsMicrosoftNXLog.json:2]: @jszigetvari-nxlog we already have custom table here : https://github.com/Azure/Azure-Sentinel/blob/master/.script/tests/KqlvalidationsTests/CustomTables/NXLog_DNS_Server_CL.json

Can you remove this file too, we don't add tables in CustomFunction?

Comment (jszigetvari-nxlog) [reply .script/tests/KqlvalidationsTests/CustomFunctions/ASimDnsMicrosoftNXLog.json]: Removed the file as you asked.

**Review by vakohl — APPROVED** [review]

**Review by v-rbajaj — APPROVED** [review]

**Pull request merged**
